tools/ci.sh

```diff
@@ -6,8 +6,6 @@ export CARGO_INCREMENTAL=0
 export CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse
 
 cargo build --bin lalrpop --features pico-args
-# build with minimal amount of features to make sure dependencies can build too.
-cargo hack build --workspace --feature-powerset --exclude-features pico-args --optional-deps
```

Comment:
Did you do any measurement on how much time dropping the feature-powerset build saves in CI?  My impression is that test is the really expensive bit.  I'm inclined to agree with dropping the powerset test.  The really important thing is just getting the single binary the tests need, but we do get *some* value here, since "cargo build" tests a slightly different configuration that "cargo test" does.  It's probably not enough value to be worth the extra CI time, but I'm curious if we have any data to compare what it's costing us to leave this in.

(If we were to leave it in, it should maybe be reordered to before the regular build to ensure we run tests with a specified lalrpop binary, rather than whatever hack happened to decide to do last.)

Reply:
This seems to hold up in that it was only a 20s saving but probably close enough to some margin of error. 

`since "cargo build" tests a slightly different configuration that "cargo test" does` In what way? Maybe this is what I'm unclear about.

Reply:
hmm running with incremental but without cargo hack build has given the best time so far... but I'm not sure what my error margin is when doing this.

Reply:
(Inspired by https://github.com/dtolnay/rust-toolchain/issues/61)

Reply:
> since "cargo build" tests a slightly different configuration that "cargo test" does In what way? Maybe this is what I'm unclear about.

IIUC there are two major differences:

1. The final binary isn't linked together.  I doubt we care too much about this difference.  I'm not sure if there's anything that potentially could go wrong here that wouldn't be a rust bug.
2.  `#[cfg(test)]` blocks wouldn't be included in build.  This could potentially cause issues if a certain block of code was conditionally only needed under certain features, but was also marked `#[cfg(test)]`, I believe.

IMHO, running just the single build without a hack, and then running test on the full hack seems like a sensible compromise between CI usage and test coverage.  I don't think we're getting a ton of added value in doing the hack build, and I worry that we could end up causing unneeded CI failures by accidentally using the wrong built version during `cargo test`.

Reply:
Yeah, I think those differences are unlikely to come up in practice. I also am in favor of removing `cargo hack build` purely on it being simpler for what is basically the same benefit(and then not needing to rely on cargo properly reusing build artifacts).